Make weird name lints trigger behind cfg_attr

est31 opened the pull request:
The weird name lints (`unknown_lints`, `renamed_and_removed_lints`), the lints that lint the linting, were previously not firing for lint level declarations behind `cfg_attr`, as they were only running before expansion.

Now, this will give a `unknown_lints` warning:

```Rust
#[cfg_attr(all(), allow(this_lint_does_not_exist))]
fn foo() {}
```

Lint level declarations behind a `cfg_attr` whose condition is not applying are still ignored. So this still won't give a warning:

```Rust
#[cfg_attr(any(), allow(this_lint_does_not_exist))]
fn foo() {}
```

Furthermore, this PR also makes the weird name lints respect level delcarations for *them* that were hidden by `cfg_attr`, making them consistent to other lints. So this will now not issue a warning:

```Rust
#[cfg_attr(all(), allow(unknown_lints))]
mod foo {
    #[allow(does_not_exist)]
    fn foo() {
    }
}
```

Fixes #97094

rust-highfive commented:
r? @lcnr

(rust-highfive has picked a reviewer for you, use r? to override)

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-12`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/6540996250?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/9e171e8786ab26e80f2d37ae879c522a1ee5abe4/checks/6540996250/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

---- [ui] src/test/ui-fulldeps/lint-tool-test.rs stdout ----
diff of stderr:

18 LL | #[allow(test_group)]
19    |         ^^^^^^^^^^ help: change it to: `clippy::test_group`
- warning: unknown lint: `this_lint_does_not_exist`
-   --> $DIR/lint-tool-test.rs:36:8
-    |
-    |
- LL | #[deny(this_lint_does_not_exist)]
-    |
-    = note: `#[warn(unknown_lints)]` on by default
- 
29 warning: lint name `test_lint` is deprecated and may not have an effect in the future.
29 warning: lint name `test_lint` is deprecated and may not have an effect in the future.
30   --> $DIR/lint-tool-test.rs:9:23
31    |

43    |
44 LL | #[allow(test_group)]
45    |         ^^^^^^^^^^ help: change it to: `clippy::test_group`
+ warning: unknown lint: `this_lint_does_not_exist`
+   --> $DIR/lint-tool-test.rs:36:8
+    |
+    |
+ LL | #[deny(this_lint_does_not_exist)]
+    |
+    = note: `#[warn(unknown_lints)]` on by default
46 
46 
47 warning: use of deprecated attribute `plugin`: compiler plugins are deprecated. See https://github.com/rust-lang/rust/pull/64675


The actual stderr differed from the expected stderr.
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui-fulldeps/lint-tool-test/lint-tool-test.stderr
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui-fulldeps/lint-tool-test/lint-tool-test.stderr
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args lint-tool-test.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/ui-fulldeps/lint-tool-test.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui-fulldeps/lint-tool-test" "-A" "unused" "-Crpath" "-O" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--cfg" "foo" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui-fulldeps/lint-tool-test/auxiliary"
stdout: none
--- stderr -------------------------------
warning: lint name `test_lint` is deprecated and may not have an effect in the future.
   |
   |
LL | #![cfg_attr(foo, warn(test_lint))]
   |                       ^^^^^^^^^ help: change it to: `clippy::test_lint`
   = note: `#[warn(renamed_and_removed_lints)]` on by default


warning: lint name `clippy_group` is deprecated and may not have an effect in the future.
   |
   |
LL | #![deny(clippy_group)]
   |         ^^^^^^^^^^^^ help: change it to: `clippy::group`

warning: lint name `test_group` is deprecated and may not have an effect in the future.
   |
   |
LL | #[allow(test_group)]
   |         ^^^^^^^^^^ help: change it to: `clippy::test_group`
warning: lint name `test_lint` is deprecated and may not have an effect in the future.
  --> /checkout/src/test/ui-fulldeps/lint-tool-test.rs:9:23
   |
   |
LL | #![cfg_attr(foo, warn(test_lint))]
   |                       ^^^^^^^^^ help: change it to: `clippy::test_lint`

warning: lint name `clippy_group` is deprecated and may not have an effect in the future.
   |
   |
LL | #![deny(clippy_group)]
   |         ^^^^^^^^^^^^ help: change it to: `clippy::group`

warning: lint name `test_group` is deprecated and may not have an effect in the future.
   |
   |
LL | #[allow(test_group)]
   |         ^^^^^^^^^^ help: change it to: `clippy::test_group`
warning: unknown lint: `this_lint_does_not_exist`
  --> /checkout/src/test/ui-fulldeps/lint-tool-test.rs:36:8
   |
   |
LL | #[deny(this_lint_does_not_exist)] //~ WARNING unknown lint: `this_lint_does_not_exist`
   |
   = note: `#[warn(unknown_lints)]` on by default


warning: use of deprecated attribute `plugin`: compiler plugins are deprecated. See https://github.com/rust-lang/rust/pull/64675
   |
   |
LL | #![plugin(lint_tool_test)]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^ help: may be removed in a future compiler version
   = note: `#[warn(deprecated)]` on by default

warning: lint name `test_lint` is deprecated and may not have an effect in the future.
  --> /checkout/src/test/ui-fulldeps/lint-tool-test.rs:9:23
  --> /checkout/src/test/ui-fulldeps/lint-tool-test.rs:9:23
   |
LL | #![cfg_attr(foo, warn(test_lint))]
   |                       ^^^^^^^^^ help: change it to: `clippy::test_lint`

warning: lint name `clippy_group` is deprecated and may not have an effect in the future.
   |
   |
LL | #![deny(clippy_group)]
   |         ^^^^^^^^^^^^ help: change it to: `clippy::group`

error: item is named 'lintme'
   |
   |
LL | fn lintme() { } //~ ERROR item is named 'lintme'
   |
note: the lint level is defined here
  --> /checkout/src/test/ui-fulldeps/lint-tool-test.rs:14:9
   |
   |
LL | #![deny(clippy_group)]
   |         ^^^^^^^^^^^^
   = note: `#[deny(clippy::test_lint)]` implied by `#[deny(clippy::group)]`

error: item is named 'lintmetoo'
   |
   |
LL |     fn lintmetoo() { } //~ ERROR item is named 'lintmetoo'
   |
   |
   = note: `#[deny(clippy::test_group)]` implied by `#[deny(clippy::group)]`

warning: lint name `test_group` is deprecated and may not have an effect in the future.
   |
   |
LL | #[allow(test_group)]
   |         ^^^^^^^^^^ help: change it to: `clippy::test_group`
warning: lint name `test_lint` is deprecated and may not have an effect in the future.
  --> /checkout/src/test/ui-fulldeps/lint-tool-test.rs:9:23
   |
   |
LL | #![cfg_attr(foo, warn(test_lint))]
   |                       ^^^^^^^^^ help: change it to: `clippy::test_lint`

warning: lint name `clippy_group` is deprecated and may not have an effect in the future.
   |
   |
LL | #![deny(clippy_group)]
   |         ^^^^^^^^^^^^ help: change it to: `clippy::group`

warning: lint name `test_group` is deprecated and may not have an effect in the future.
   |
   |
LL | #[allow(test_group)]
   |         ^^^^^^^^^^ help: change it to: `clippy::test_group`
error: aborting due to 2 previous errors; 14 warnings emitted
------------------------------------------


```

</details>

lcnr commented:
Do we want that, considering that in the future one might write `#[cfg_attr(version = "1.60", allow(some_new_lint)]`? Or am I misunderstanding something here

est31 commented:
@lcnr it will only lint if the condition inside the cfg evaluates to true. If the condition is false, it won't lint. See the last entry in the added test. I've also edited the PR description to make this clearer. Making it peek inside all `cfg_attr` regardless of whether they apply or not is a possibility, but I agree that it's not a good idea, which is why I didn't chose it for this PR.

lcnr commented on compiler/rustc_lint/src/early.rs:
```diff
@@ -446,12 +446,13 @@ pub fn check_ast_node<'a>(
                     early_lint_node(
                         sess,
-                        pre_expansion && i == 0,
+                        !pre_expansion && i == pass_count,
```
when is `i == pass_count` true? `i` should be in the range `0..pass_count` :thinking:

also, can you add a comment on why that `i == X` check is needed?

est31 replied:
Good catch! The test works great locally but I suppose the other condition of the if is visited.

est31 replied:
Indeed `i == 0` works too, I've tried it. I've also added a revision to the test file so that it's actually being tested.

lcnr commented on compiler/rustc_lint/src/early.rs:
```diff
@@ -426,7 +426,7 @@ pub fn check_ast_node<'a>(
     if !sess.opts.debugging_opts.no_interleave_lints {
```
not part of this PR, but can you swap these two branches so that the `if` has a positive condition?

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/6555011017?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/ab6357bf9645d98bbd5234d4981c84906a0b866d/checks/6555011017/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
    Checking rustc_mir_build v0.0.0 (/checkout/compiler/rustc_mir_build)
error[E0425]: cannot find value `i` in this scope
   --> compiler/rustc_lint/src/early.rs:431:23
    |
431 |         let is_last = i == (pass_count - 1);
    |
   ::: /checkout/compiler/rustc_ast/src/ptr.rs:39:1
    |
    |
39  | pub fn P<T: 'static>(value: T) -> P<T> {
    | -------------------------------------- similarly named function `P` defined here
help: a function with a similar name exists
    |
    |
431 |         let is_last = P == (pass_count - 1);
help: consider importing this constant
    |
17  | use rustc_span::sym::i;
    |
```

</details>

est31 commented:
I'm now using `i == 0` because it still works, and have changed the test to ensure that it works with that debug flag enabled. I've also reordered the bodies as requested. re-r? @lcnr

lcnr commented on compiler/rustc_lint/src/early.rs:
```diff
@@ -424,10 +424,25 @@ pub fn check_ast_node<'a>(
+                sess.prof.extra_verbose_generic_activity("run_lint", pass.name()).run(|| {
+                    early_lint_node(
+                        sess,
+                        !pre_expansion && i == 0,
```
```suggestion
                        // We want to only warn about weird lints
                        // once. As it doesn't matter when we lint,
                        // we do it as part of the first pass.
                        !pre_expansion && i == 0,
```
That's my understanding here, is that correct?

lcnr reviewed (commented):
one nit, then r=me

est31 commented:
@lcnr your understanding is correct, yes. To make sure, I've removed the `== 0` check locally and tested it, and indeed it emitted the warnings multiple times, so the `== 0` is needed. I've added your suggestion to the PR.

lcnr commented on src/test/ui/lint/issue-97094.rs:
```diff
@@ -0,0 +1,51 @@
+);
+
+// Allowing unknown lints works if inside cfg_attr
+#[cfg_attr(all(), allow(unknown_lints))]
```
what happens for
```rust
#[cfg_attr(any(), allow(unknown_lints))]
mod bar {
    #[cfg_attr(all(), allow(nonex_lint_fn))]
    fn _foo() {}
}
``` does that also silence the lint for `nonex_lint_fn`?

lcnr replied:
ah damn, that's cool

lcnr commented:
@bors r+

bors commented:
:pushpin: Commit 7a36d24d93d00569c7907228f96d259f38a6f571 has been approved by `lcnr`

<!-- @bors r=lcnr 7a36d24d93d00569c7907228f96d259f38a6f571 -->
<!-- homu: {"type":"Approved","sha":"7a36d24d93d00569c7907228f96d259f38a6f571","approver":"lcnr"} -->

est31 commented:
Sorry for pinging you again, @lcnr I've improved the last two tests a little to not have a `cfg_attr` inside the module but only a `cfg`. Now they read `#[allow(nonex_lint_fn)]`. I think this improves test readability because the thing being tested in those two last tests is the `cfg_attr` of the module outside. re-r? @lcnr

lcnr commented:
@bors r+

bors commented:
:pushpin: Commit 2a8b60f915c9fdc61e18be3392ea4330d9617da4 has been approved by `lcnr`

<!-- @bors r=lcnr 2a8b60f915c9fdc61e18be3392ea4330d9617da4 -->
<!-- homu: {"type":"Approved","sha":"2a8b60f915c9fdc61e18be3392ea4330d9617da4","approver":"lcnr"} -->
